### ISSUE TITLE
generateSyntheticData effect size

### ISSUE THREAD

**Issue by dsrobertson** [csoneson/compcodeR#12]
Thanks so much for this package!

I'm having some trouble understanding the `effect.size` argument in the `generateSyntheticData` function. As seen in the example below, setting `effect.size = 3` seems to correspond to the upregulated genes in condition S2 having mean counts only 1.7 times as large as those genes in condition S1, whereas from the documentation for `effect.size` I was expecting the mean counts to be ~4 times as large.

Am I misunderstanding the documentation? Apologies if this is a naïve question, I'm not very familiar with RNA-seq data.

```r
library(compcodeR)

set.seed(1)

B_625_625 <- generateSyntheticData(dataset = "B_625_625", n.vars = 12500, 
                                   samples.per.cond = 5, n.diffexp = 6000, 
                                   repl.id = 1, seqdepth = 1e7, 
                                   fraction.upregulated = 1,
                                   effect.size = 3,
                                   between.group.diffdisp = FALSE, 
                                   filter.threshold.total = 1, 
                                   filter.threshold.mediancpm = 0, 
                                   fraction.non.overdispersed = 0, 
                                   output.file = "B_625_625_5spc_repl1.rds")


upregulated <- B_625_625@variable.annotations$upregulation
condition <- B_625_625@sample.annotations$condition

mean(B_625_625@count.matrix[upregulated == 1, condition == 2])/
  mean(B_625_625@count.matrix[upregulated == 1, condition == 1])
# 1.737832
```

**Comment by csoneson** [comment]
Thanks for the question! I think there are several things at play here: 

* The total count is not constant across samples, so the counts are not directly comparable, and thus calculating the mean of the counts directly can be misleading (the sequencing depth for each sample is obtained by multiplying the `seqdepth` by a uniformly drawn value between `minfact` (by default 0.7) and `maxfact` (by default 1.4) - see: https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L140-L141
* You are simulating a _lot_ of differentially expressed genes, and they are all regulated in the same direction (`fraction.upregulated` = 1). This type of situations are prone to cause problems for, e.g., typical RNA-seq normalization methods, which assume that most genes don't change expression between samples (which is often a reasonable assumption). 
* What's happening internally is that the "base means" in condition 1 are sampled from the example data, and then the base means of the genes in condition 2 are obtained by multiplying or dividing the condition 1 base means by the effect size for that gene. Before generating the counts, however, these "base means" are divided by the sum of the base means across all genes (which will attenuate the signal if you have a very large fraction of the genes all changing in the same direction), and multiplied by the sequencing depth for the sample, see https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L155-L166 and https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L200-L218

The "M values" included in the output object (`variable.annotations$M.value`) provide normalized count log-ratios estimated from the data, which are similar in spirit to your last few lines above, but account for the differences in sequencing depth as well as any compositional changes, see: https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L284-L290 
The `truelog2foldchanges` column, in turn, give the true underlying log2 fold changes (log ratio between the base means in the two groups): https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L170 
You will see that for upregulated genes, the lower bound of the latter is `log2(effect.size)` (since the gene-wise effect size is obtained by taking the provided effect size and adding a random value drawn from an exponential distribution). The M values and true log2 fold changes often show a high correlation, as expected (there is a plot generated by `summarizeSyntheticDataSet()` relating the two). 

So to summarize, the `effect.size` you provide becomes a lower bound for the true ratios between the base means in the two groups, which are each in turn normalized to create "relative expression values" for each condition, and multiplied by the sequencing depth for each sample before sampling the final counts.

**Comment by latlio** [comment]
Hi Charlotte,

Thanks for your detailed response. 

Given that `fraction.upregulated = 1` is "prone to cause problems for, e.g., typical RNA-seq normalization methods", I was wondering why that's set as the default value, or did I misunderstand you here.

Also, I was wondering if I'm misunderstanding something else about this function. We want to simulate different RNAseq datasets with various proportion of DE genes (specified by `pi_vec` below). However, what I find is that the generated counts do not look different between the different proportions. You mentioned that "the M values and true log2 fold changes often show a high correlation", so I was surprised to see the lack of change between proportions in the count histograms. Hopefully this reprex can run on your machine.

Thanks!

```
library(tidyverse)
library(cowplot)
library(compcodeR)

# Set Parameters
n_genes <- 5000
n_samples <- 5
pi_vec <- round(c(seq(0.01, 0.09, by = 0.01), seq(0.1, 0.5, by = 0.1)), 2)
my_fc <- 2
effect_size_list <- list()
sim_counts <- list()

create_per_gene_effect_sizes <- function(n_genes, pi, fc) {
  c(rep(2^fc, n_genes*pi), rep(1, n_genes-(n_genes*pi)))
}

for(i in seq_along(1:length(pi_vec))) {
  effect_size_list[[i]] <- tibble(effect_size = create_per_gene_effect_sizes(n_in_batch, pi_vec[i], my_fc),
                                   id = paste0("g", seq(n_in_batch))) %>%
    mutate(truth = ifelse(effect_size == 2^my_fc, 1, 0))
  
  sim_counts2[[i]] <- generateSyntheticData(dataset = paste0("sim", i, "-", pi_vec[i]),
                                            n.vars = n_in_batch,
                                            samples.per.cond = n_samples,
                                            repl.id = i,
                                            minfact = 0.9,
                                            maxfact = 1.1,
                                            fraction.upregulated = 0.5,
                                            effect.size = effect_size_list[[i]]$effect_size)
}

raw_experimental_count_histograms <- imap(sim_counts, ~ .x@count.matrix %>%
                                             as.data.frame() %>%
                                             select(6:10) %>% #experimental samples
                                             pivot_longer(everything(), 
                                                          names_to = "sample",
                                                          values_to = "count") %>%
                                             ggplot(aes(x = count)) + 
                                             geom_histogram(bins = 50) + 
                                             theme_bw() + 
                                             labs(title = .y))
cowplot::plot_grid(plotlist = raw_experimental_count_histograms)
```

**Comment by csoneson** [comment]
> Given that fraction.upregulated = 1 is "prone to cause problems for, e.g., typical RNA-seq normalization methods", I was wondering why that's set as the default value, or did I misunderstand you here.

That is a fair point, and if I would have released the function now, I would probably have set it to 0.5. I have been hesitant to change the default value to avoid the output suddenly changing for users. It is not really a big problem unless a large fraction of the genes are differentially expressed (see e.g. panels A-D [here](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91/figures/4)). 

> However, what I find is that the generated counts do not look different between the different proportions. You mentioned that "the M values and true log2 fold changes often show a high correlation", so I was surprised to see the lack of change between proportions in the count histograms.

The distribution of the counts within a specific condition are not necessarily expected to be different depending on the number of DE genes. If you look at the distribution of M-values, you see a clear difference: 

```
cowplot::plot_grid(plotlist = lapply(sim_counts, function(x) {
    ggplot(x@variable.annotations, aes(x = M.value)) + geom_histogram(bins = 100)
}))
```

**Comment by latlio** [comment]
> The distribution of the counts within a specific condition are not necessarily expected to be different depending on the number of DE genes. If you look at the distribution of M-values, you see a clear difference

Thanks, this was a helpful distinction.

As a follow-up, I feel like I'm still running into issues for how to specify downregulated genes. If the target is to simulate 2500 DE genes (and 2500 non-DE genes), with 1250 of those DE being upregulated, and 1250 being downregulated (i.e. `fraction.upregulated = 0.5`), then my understanding of your function leads me to this:

```
test <- generateSyntheticData(dataset = "B_2500_2500",
                              n.vars = 5000,
                              samples.per.cond = 5,
                              repl.id = 1,
                              minfact = 0.9,
                              maxfact = 1.1,
                              effect.size = c(rep(3, 1250), rep(-3, 1250), rep(1, 2500)))
```

where the effect sizes are explicitly specified as up and down by + and - respectively.

https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L93-L98

However, I run into the following error: Error in calcNormFactors.default(Z) : NA counts not permitted. (Hopefully you can reproduce this)

If I want to specify a vector for `effect.size` in this way, how do I go about it without encountering the `calcNormFactors` error?

**Comment by csoneson** [comment]
Note that the `effect.size` specifies multiplicative effects: https://github.com/csoneson/compcodeR/blob/7d50e6985a063b72bcf524cdbd9e8935afdd86a8/R/generateSyntheticData.R#L168
So you want positive values > 1 for upregulated genes, positive values < 1 for downregulated ones, and values = 1 for non-DE genes.

**Comment by latlio** [comment]
Ohh, thank you!

**Comment by dsrobertson** [comment]
Just to add to what @latlio has said (we're working together on the same research project) - your replies have been most helpful! Thanks so much!
